src/Servers/Kestrel/Core/src/KestrelConfigurationLoader.cs

```diff
@@ -429,20 +430,133 @@ private bool TryGetCertificatePath(out string path)
 
         private X509Certificate2 LoadCertificate(CertificateConfig certInfo, string endpointName)
         {
+            var logger = Options.ApplicationServices.GetRequiredService<ILogger<KestrelConfigurationLoader>>();
             if (certInfo.IsFileCert && certInfo.IsStoreCert)
             {
                 throw new InvalidOperationException(CoreStrings.FormatMultipleCertificateSources(endpointName));
             }
             else if (certInfo.IsFileCert)
             {
-                var env = Options.ApplicationServices.GetRequiredService<IHostEnvironment>();
-                return new X509Certificate2(Path.Combine(env.ContentRootPath, certInfo.Path), certInfo.Password);
+                var environment = Options.ApplicationServices.GetRequiredService<IHostEnvironment>();
+                var certificatePath = Path.Combine(environment.ContentRootPath, certInfo.Path);
+                if (certInfo.KeyPath != null)
+                {
+                    var certificateKeyPath = Path.Combine(environment.ContentRootPath, certInfo.KeyPath);
+                    var certificate = GetCertificate(certificatePath);
+
+                    if (certificate != null)
+                    {
+                        certificate = LoadCertificateKey(certificate, certificateKeyPath, certInfo.Password);
+                    }
+                    else
+                    {
+                        logger.FailedToLoadCertificate(certificateKeyPath);
+                    }
+
+                    if (certificate != null)
+                    {
+                        if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+                        {
+                            return PersistKey(certificate);
+                        }
+
+                        return certificate;
+                    }
+                    else
+                    {
+                        logger.FailedToLoadCertificateKey(certificateKeyPath);
+                    }
+
+                    throw new InvalidOperationException(CoreStrings.InvalidPemKey);
```

Comment:
This could throw if the cert is missing, not just because the key was missing or invalid right? This exception message implies the key must be the problem?

And why do we use the exact same exception and log message any time the key is missing or invalid? It would be a lot better to log exactly why the key is invalid and to be very clear when the key is actually missing vs being invalid in both the exception and log messages.